src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs

```diff
@@ -166,6 +171,12 @@ private void SetCurrentPruning(IPruningContext pruningContext)
 
         private bool CanStartNewPruning() => _fullPruningDb.CanStartPruning;
 
+        private void HandlePruningFinished(object? sender, PruningEventArgs e)
+        {
+            if (_logger.IsInfo) _logger.Info("Full Pruning completed, shutting down as requested in the configuration.");
+            Environment.Exit(0);
```

Comment:
Not sure if optimal way of shutting down. Attached logs potentially show that not all things were disposed properly (like DB's), not sure if its only logs+docker thing or they wern't disposed. Potentially better to do by reusing this code: https://github.com/NethermindEth/nethermind/blob/a3d91a5f159476cec9ccb43606cedc67773b4418/src/Nethermind/Nethermind.Runner/Program.cs#L452-L457

Reply:
I looked through, and you hook both a Console cancel and a general Application Exit here:

https://github.com/NethermindEth/nethermind/blob/a3d91a5f159476cec9ccb43606cedc67773b4418/src/Nethermind/Nethermind.Runner/Program.cs#L395-L399

The intent was to activate that second hook, but if it's not working as intended, then I can trigger the first one. I'll add some stuff so FullPruner can access that cancel source in a clean way.

Reply:
This actually triggers a graceful shutdown as expected, but Nethermind is interpreting it in a way that prevents the shutdown from working properly - see https://github.com/NethermindEth/nethermind/issues/4057 for another example of this behavior. The ideal solution would be to fix whatever is causing the anomaly with SIGTERM, but if that isn't feasible, this could be changed to emit a SIGINT instead.

Reply:
Can you change it to SIGTERM now if the shutdown isn't graceful, and leave a comment for the future?

Reply:
Essentially, what it's doing now already is SIGTERM. `Environment.Exit(0)` is behaving exactly the same way a Docker shutdown request behaves, so from that perspective this is working as intended and doesn't need any adjustment. My proposal was to change it to "nonstandard behavior" (something that emits SIGINT instead) which correctly shuts things down, and put a note in there requesting someone to change it back to `Environment.Exit(0)` once the SIGTERM issue has been resolved in Nethermind as a whole.

Reply:
Ok can you change it to SIGINT for now?

Reply:
This should fix sigterm issue https://github.com/NethermindEth/nethermind/pull/4163. Ideally you would want to pass in the cancellation source. Which one though, I'm not sure, sorry. If you can confirm that the `All done, goodbye!` message is shown, I'm good with this.

Reply:
> This should fix sigterm issue #4163. Ideally you would want to pass in the cancellation source. Which one though, I'm not sure, sorry. If you can confirm that the `All done, goodbye!` message is shown, I'm good with this.

I can confirm that your fix works - it now behaves properly with `SIGTERM`. I ran a full prune on Kiln as a quick test and found that it ran into a concurrency issue, so I moved the `Environment.Exit()` call into a separate Task. This is now the behavior:

```
2022-06-23 04:40:24.2997|Full Pruning Finished: 00:00:03.4428119 0.21 mln nodes mirrored.
...
2022-06-23 04:40:26.2985|Disposing DB State0 
2022-06-23 04:40:26.3552|Full Pruning completed, shutting down as requested in the configuration. 
2022-06-23 04:40:26.3552|Closing, please wait until all functions are stopped properly... 
...
(Lots of closing / disposing messages)
...
2022-06-23 04:40:31.8104|Disposing DB CanonicalHashTrie 
2022-06-23 04:40:31.8108|Disposing DB Code 
2022-06-23 04:40:31.8108|All DBs closed. 
2022-06-23 04:40:31.8108|Ethereum shutdown complete... please wait for all components to close 
2022-06-23 04:40:31.8108|All done, goodbye! 
```

The `All done, goodbye!` message appears, which implies this is shutting down properly. Well done!